src/lib.rs

```diff
@@ -65,90 +84,183 @@ pub(crate) fn generate_seed(initial_seed: &[u8]) -> Result<[u8; 32], &str> {
     Ok(seed)
 }
 
+// Decode from a base-64 JWS string into a JWS helper struct
+pub fn decode_jws(jws_b64: &str) -> Result<JWS, Error> {
+    let mut itr = jws_b64.splitn(3, ".").map(|slice| {
+        base64::decode_config(slice, base64::URL_SAFE_NO_PAD).unwrap()
+    });
+
+    let (header, payload, signature) = match (itr.next(), itr.next(), itr.next()) {
+        (Some(h), Some(p), Some(s)) => (h, p, s),
+        _ => return Err(Error::DecodeError)
+    };
+
+    Ok(
+        JWS {
+            header: serde_json::from_slice(&header).unwrap(),
+            payload: payload.to_vec(),
+            signature: signature.to_vec(),
+        }
+    )
+}
+
+// Translate a JWS helper struct into the IetfJsonPatch serializable struct
+pub fn get_json_patch(jws: &JWS) -> Result<IetfJsonPatch, Error> {
+    match serde_json::from_slice::<IetfJsonPatch>(&jws.payload) {
+        Ok(docs) => Ok(docs),
+        Err(_) => Err(Error::DecodeError)
+    }
+}
+
+// Per the spec (https://bit.ly/34xScAu), verify JWS patch change is from the controller
+pub fn verify_json_patch_jws(jws: &JWS, key: &KeyPair) -> bool {
+    let kid = &jws.header.key_id;
+    let did = key.get_did_document(Config::default());
+
+    if let Some(authentication) = did.authentication {
+        if let Some(_key_id) = authentication.iter().find(|&s| *s == kid.clone().unwrap()) {
+            match key.verify(&jws.payload, &jws.signature) {
+                Ok(()) => { return true; },
+                Err(_) => { return false; },
+            }
+        }
+    }
+    return false;
+}
+
+// Use json_patch helpers to update a DID JSON document. Upon error, return original document.
+pub fn patch_json_document(doc: &Document, json_patch: &IetfJsonPatch) -> Document {
+    let original = doc.clone();
+
+    fn apply_patch(doc: &Document, json_patch: &IetfJsonPatch) -> Result<Document, Box<dyn std::error::Error>> {
+        let parsed_patch = from_value(json_patch.value.clone())?;
+        let mut json_doc = serde_json::to_value(doc.clone())?;
+        patch(&mut json_doc, &parsed_patch)?;
+        serde_json::from_value(json_doc)
+            .or_else(|err| Err(Box::new(err) as Box<dyn std::error::Error>))
```

Comment:
this ugly syntax is to appease the compiler, which *should* have been able to infer type from `?` but apparently has issues with `serde_json` errors like this [[ref](https://stackoverflow.com/questions/57423880/handling-serde-error-and-other-error-type)].  

my goal here is to wrap up all error types together (using `Box`) and simply return the original in that case.  if anyone has thoughts on how to achieve this in a better/prettier way, I am all ears!

Reply:
Not here, no - the issue is that `serde_json` errors can't be automatically interpreted as a generic `dyn std::error::Error` because "the trait `Deserialize<'_>` is not implemented for `dyn StdError`".  

I did realize I can use cleaner syntax with `.into()` though, so I'll update with that!

Reply:
Agreed that implementing `Into<std::error::Error>` is often a good idea for custom error types.  I don't see a reason to need it yet (only error-boxing is here, where we box any generic non-custom error type into this `Err` arm to return the original doc).  But likely we'd want it later